File: go/src/421.数组中两个数的最大异或值.go
/*
 * @lc app=leetcode.cn id=421 lang=golang
 *
 * [421] 数组中两个数的最大异或值
 */

// @lc code=start
var used int
var trie [][]int

const TRIE_LEN = 1e6

func init() {
	trie = make([][]int, TRIE_LEN)
	for i := 0; i < TRIE_LEN; i++ {
		trie[i] = make([]int, 2)
	}
}

func findMaximumXOR(nums []int) int {
	used = 1
	for i := 0; i < TRIE_LEN; i++ {
		trie[i][0], trie[i][1] = 0, 0
	}

	res := 0
	for _, num := range nums {
		add(trie, num)
		res = max(res, findMax(trie, num))
	}

	return res
}

func add(trie [][]int, num int) {
	index := 0
	for i := 30; i >= 0; i-- {
		a := (num >> i) & 1
		if trie[index][a] == 0 {
			trie[index][a] = used
			used++
		}
		index = trie[index][a]
	}
}

func findMax(trie [][]int, num int) int {
	res, index := 0, 0
	for i := 30; i >= 0; i-- {
		a := (num >> i) & 1
		b := 1 - a
		if trie[index][b] != 0 {
			res |= 1 << i
			index = trie[index][b]
		} else {
			index = trie[index][a]
		}
	}

	return res
}

func max(a ...int) int {
	ans := a[0]
	for _, v := range a {
		if v > ans {
			ans = v
		}
	}
	return ans
}

// @lc code=end

File: rust/src/421.数组中两个数的最大异或值.rs
/*
 * @lc app=leetcode.cn id=421 lang=rust
 *
 * [421] 数组中两个数的最大异或值
 */

// @lc code=start
use std::collections::HashSet;

impl Solution {
    pub fn find_maximum_xor(nums: Vec<i32>) -> i32 {
        let mut res = 0;
        let mut mask = 0;
        let mx = *nums.iter().max().unwrap();
        let high_bit = 31 - mx.leading_zeros() as i32;

        let mut seen = HashSet::new();

        for i in (0..=high_bit).rev() {
            seen.clear();
            mask |= 1 << i;
            let new_res = res | (1 << i);
            for &x in &nums {
                let x = x & mask;
                if seen.contains(&(x ^ new_res)) {
                    res = new_res;
                    break;
                }
                seen.insert(x);
            }
        }

        res
    }
}
// @lc code=end
